base/src/main/java/uk/ac/ebi/atlas/model/experiment/ExperimentDesign.java

```diff
@@ -150,46 +141,39 @@ public void addAssayHeader(String assayHeader) {
         return null;
     }
 
-
     public ImmutableSetMultimap<String, String> getAllOntologyTermIdsByAssayAccession() {
         ImmutableSetMultimap.Builder<String, String> builder = ImmutableSetMultimap.builder();
 
-        addFactorOntologyTerms(builder);
-        addCharacteristicOntologyTerms(builder);
+        getFactorOntologyTerms().forEach(builder::putAll);
+        getCharacteristicOntologyTerms().forEach(builder::putAll);
 
         return builder.build();
     }
 
-
-    private void addCharacteristicOntologyTerms(ImmutableSetMultimap.Builder<String, String> builder) {
-        for (Map.Entry<String, SampleCharacteristics> sampleEntry : samples.entrySet()) {
-            String runOrAssay = sampleEntry.getKey();
-            SampleCharacteristics sampleCharacteristics = sampleEntry.getValue();
-
-            for (SampleCharacteristic sampleCharacteristic : sampleCharacteristics.values()) {
-                for (OntologyTerm valueOntologyTerm : sampleCharacteristic.valueOntologyTerms()) {
-                    builder.put(runOrAssay, valueOntologyTerm.accession());
-                }
-            }
-
-        }
+    private Map<String, Set<String>> getFactorOntologyTerms() {
+        return factorSetMap.entrySet().stream()
+                .collect(toMap(
+                        Map.Entry::getKey,
+                        entry ->
+                                stream(entry.getValue().spliterator(), false)
+                                        .flatMap(factor ->
+                                                factor.getValueOntologyTerms().stream())
+                                        .map(OntologyTerm::accession)
+                                        .collect(toSet())));
     }
 
-
-    private void addFactorOntologyTerms(ImmutableSetMultimap.Builder<String, String> builder) {
-        for (Map.Entry<String, FactorSet> factorSetEntry : factorSetMap.entrySet()) {
-            String runOrAssay = factorSetEntry.getKey();
-            FactorSet factorSet = factorSetEntry.getValue();
-
-            for (Factor factor : factorSet) {
-                for (OntologyTerm valueOntologyTerm : factor.getValueOntologyTerms()) {
-                    builder.put(runOrAssay, valueOntologyTerm.accession());
-                }
-            }
-        }
+    private Map<String, Set<String>> getCharacteristicOntologyTerms() {
+        return samples.entrySet().stream()
+                .collect(toMap(
+                        Map.Entry::getKey,
+                        entry ->
+                                entry.getValue().values().stream()
+                                        .flatMap(sampleCharacteristic ->
+                                                sampleCharacteristic.valueOntologyTerms().stream())
```

Comment:
Due to `SampleCharacteristic` using `AutoValue`, it means that the naming convention is ever so slightly different. You have `sampleCharacteristic.valueOntologyTerms()` but you have `factor.*get*ValueOntologyTerms()`. Besides that these getters are pretty much identical... Some inheritance and refactoring the models might simplify things a bit.

Reply:
Ahh okay, I just noticed the `FactorSet` complication. Nevermind...